teps/0022-trigger-immutable-input.md

```diff
@@ -0,0 +1,351 @@
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation.  The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### Webhook Interceptors
+
+#### Input Event Field
+
+Instead of mixing input event fields with others (e.g. the `webhook.header`
+field) we will include the entire input event in a reserved body field named
+`input`. This field will not be mutated throughout trigger processing, so that
+inteceptors and trigger bindings can rely on assumptions on the incoming event
+type. Incoming event data will be stored within `input` as `body` and `headers`
+fields.
+
+##### Example
+
+Before:
+
+```
+POST /interceptor HTTP/1.1
+Host: my-interceptor:4567
+X-GitHub-Delivery: 72d3162e-cc78-11e3-81ab-4c9367dc0958
+X-Hub-Signature: sha1=7d38cdd689735b008b3c702edd92eea23791c5f6
+User-Agent: GitHub-Hookshot/044aadd
+Content-Type: application/json
+Content-Length: 6615
+X-GitHub-Event: issues
+Tekton-Event-ID: abc123
+{
+  "action": "opened",
+  "issue": {
+    "url": "https://api.github.com/repos/octocat/Hello-World/issues/1347",
+    "number": 1347,
+    ...
+  },
+  "repository" : {
+    "id": 1296269,
+    "full_name": "octocat/Hello-World",
+    "owner": {
+      "login": "octocat",
+      "id": 1,
+      ...
+    },
+    ...
+  },
+  "sender": {
+    "login": "octocat",
+    "id": 1,
+    ...
+  }
+}
+```
+
+After:
+
+```
+POST /interceptor HTTP/1.1
+Host: my-interceptor:4567
+Tekton-Event-ID: abc123
+{
+  "input": {
+    "headers": [
+      "X-GitHub-Delivery": "72d3162e-cc78-11e3-81ab-4c9367dc0958",
+      "X-Hub-Signature": "sha1=7d38cdd689735b008b3c702edd92eea23791c5f6",
+      "User-Agent": "GitHub-Hookshot/044aadd",
+      "Content-Type": "application/json",
+      "Content-Length": "6615",
+      "X-GitHub-Event": "issues",
+    ],
+    "body": {
+      "action": "opened",
+      "issue": {
+        "url": "https://api.github.com/repos/octocat/Hello-World/issues/1347",
+        "number": 1347,
+        ...
+      },
+      "repository" : {
+        "id": 1296269,
+        "full_name": "octocat/Hello-World",
+        "owner": {
+          "login": "octocat",
+          "id": 1,
+          ...
+        },
+        ...
+      },
+      "sender": {
+        "login": "octocat",
+        "id": 1,
+        ...
+      }
+    }
+  }
+}
+```
+
+#### Interceptor Field
+
+We propose that any responses returned from interceptors are now additive to the
+interceptor chain payload, and remove the ability to arbitrarily modify
+payloads.
+
+Any response received from interceptors will be included in following paylods
+under a new `interceptors` field, keyed by interceptor name for uniqueness.
+Webhook interceptors may no longer modify headers as part of the interceptor
+chain.
+
+If no response body is returned by the interceptor, no `interceptors.<name>`
```

Comment:
This might complicate bindings, if you have several interceptors, and the values go in as...

```json
{
  "interceptors": {
    "github-push-repo-a-interceptor": {
      "short_sha": "bca6b70"
    },
    "github-push-repo-b-interceptor": {
      "short_sha": "3e29f82b"
    }
  }
}
```
Getting a binding that can build and tag with the `short_sha` in a generic form will be tricky.

Reply:
The scope for interceptors to do this is fairly small, there's no "global" interceptor, so if it's happening, it's happening within a _fairly_ small stanza in your YAML.

To me, this suggests a low likelihood of having to debug a mysterious value, and with "last-one-wins", it should be fairly easy to debug (it's not like we're subject to the vagaries of hash ordering).